File: memories/2024/exercise/java/day13/src/main/java/santamarket/model/ChristmasElf.java
package santamarket.model;

import java.util.*;

public class ChristmasElf {
    private final SantamarketCatalog catalog;
    private final Map<Product, Offer> offers = new HashMap<>();

    public ChristmasElf(SantamarketCatalog catalog) {
        this.catalog = catalog;
    }

    public void addSpecialOffer(SpecialOfferType offerType, Product product, double argument) {
        offers.put(product, new Offer(offerType, product, argument));
    }

    public Receipt checksOutArticlesFrom(ShoppingSleigh sleigh) {
        Receipt receipt = new Receipt();
        
        for (ProductQuantity item : sleigh.getItems()) {
            Product p = item.getProduct();
            double quantity = item.getQuantity();
            double unitPrice = catalog.getUnitPrice(p);
            receipt.addProduct(quantity * unitPrice);
        }
        
        sleigh.handleOffers(receipt, offers, catalog);
        
        return receipt;
    }
}

File: memories/2024/exercise/kotlin/day13/src/main/kotlin/santamarket/model/ChristmasElf.kt
package santamarket.model

class ChristmasElf(private val catalog: SantamarketCatalog) {
    private val offers = mutableMapOf<Product, Offer>()

    fun addSpecialOffer(offerType: SpecialOfferType, product: Product, argument: Double) {
        offers[product] = Offer(offerType, product, argument)
    }

    fun checksOutArticlesFrom(sleigh: ShoppingSleigh): Receipt {
        val receipt = Receipt()

        for (item in sleigh.getItems()) {
            val p = item.product
            val quantity = item.quantity
            val unitPrice = catalog.getUnitPrice(p)
            receipt.addProduct(quantity * unitPrice)
        }

        sleigh.handleOffers(receipt, offers, catalog)

        return receipt
    }
}
